pkg/driver/directory_provisioner.go

```diff
@@ -0,0 +1,135 @@
+			return nil, status.Errorf(codes.InvalidArgument, "Parameter %v cannot be empty", FsId)
+		}
+		fileSystemId = value
+	} else {
+		return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", FsId)
+	}
+	klog.V(5).Infof("Provisioning directory on FileSystem %s...", fileSystemId)
+
+	localCloud, roleArn, err := getCloud(d.cloud, req.GetSecrets())
+	if err != nil {
+		return nil, err
+	}
+
+	mountOptions, err := getMountOptions(ctx, localCloud, fileSystemId, roleArn)
+	if err != nil {
+		return nil, err
+	}
+	target := TempMountPathPrefix + "/" + uuid.New().String()
+	if err := d.mounter.MakeDir(target); err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not create dir %q: %v", target, err)
+	}
+	if err := d.mounter.Mount(fileSystemId, target, "efs", mountOptions); err == nil {
+		// Extract the basePath
+		var basePath string
+		if value, ok := volumeParams[BasePath]; ok {
+			basePath = value
+		}
+
+		rootDirName := req.Name
+		provisionedPath = basePath + "/" + rootDirName
+
+		klog.V(5).Infof("Provisioning directory at path %s", provisionedPath)
+
+		// Grab the required permissions
+		perms := os.FileMode(0755)
+		if value, ok := volumeParams[DirectoryPerms]; ok {
+			parsedPerms, err := strconv.ParseUint(value, 8, 32)
+			if err == nil {
+				perms = os.FileMode(parsedPerms)
+			}
+		}
+
+		klog.V(5).Infof("Provisioning directory with permissions %s", perms)
+
+		provisionedDirectory := path.Join(target, provisionedPath)
+		err := d.osClient.MkDirAllWithPerms(provisionedDirectory, perms, uid, gid)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not provision directory: %v", err)
+		}
+	} else {
+		return nil, status.Errorf(codes.Internal, "Could not mount %q at %q: %v", fileSystemId, target, err)
+	}
+
+	err = d.mounter.Unmount(target)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not unmount %q: %v", target, err)
+	}
+	err = d.osClient.RemoveAll(target)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not delete %q: %v", target, err)
+	}
+
+	return &csi.Volume{
+		CapacityBytes: req.GetCapacityRange().GetRequiredBytes(),
+		VolumeId:      fileSystemId + ":" + provisionedPath,
+		VolumeContext: map[string]string{},
+	}, nil
+}
+
+func (d DirectoryProvisioner) Delete(ctx context.Context, req *csi.DeleteVolumeRequest) error {
+	if d.deleteProvisionedDir {
+		fileSystemId, subpath, _, _ := parseVolumeId(req.GetVolumeId())
+
+		localCloud, roleArn, err := getCloud(d.cloud, req.GetSecrets())
+		if err != nil {
+			return err
+		}
+
+		mountOptions, err := getMountOptions(ctx, localCloud, fileSystemId, roleArn)
+		if err != nil {
+			return err
+		}
+
+		target := TempMountPathPrefix + "/" + uuid.New().String()
+		if err := d.mounter.MakeDir(target); err != nil {
+			return status.Errorf(codes.Internal, "Could not create dir %q: %v", target, err)
+		}
+		if err := d.mounter.Mount(fileSystemId, target, "efs", mountOptions); err != nil {
```

Comment:
I did this is two separate defer statements so that each part had a chance to run at least even if one part failed, but I'm happy to change it if that's not what you meant